date-time/src/commonTest/kotlin/com/splendo/kaluga/datetime/timer/MonotonicTimerTest.kt

```diff
@@ -0,0 +1,119 @@
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.datetime.timer
+
+import com.splendo.kaluga.base.runBlocking
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withTimeout
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertIs
+import kotlin.test.assertTrue
+import kotlin.time.Duration
+
+class MonotonicTimerTest {
+
+    private fun checkStateTransitions(tickCorrection: TickCorrection): Unit = runBlocking {
+        val timer = TimerProvider.monotonic(
+            duration = Duration.milliseconds(100),
+            interval = Duration.milliseconds(10),
+            tickCorrection = tickCorrection,
+            coroutineScope = this,
+            coroutineContext = this.coroutineContext
+        )
+        assertIs<Timer.State.NotRunning.Paused>(timer.state.value, "timer was not paused after creation")
+        timer.start()
+        assertIs<Timer.State.Running>(timer.state.value, "timer is not running after start")
+        timer.pause()
+        assertIs<Timer.State.NotRunning.Paused>(timer.state.value, "timer was not paused after pause")
+        delay(200)
+        assertIs<Timer.State.NotRunning.Paused>(timer.state.value, "timer pause is not working")
+        timer.start()
+        assertIs<Timer.State.Running>(timer.state.value, "timer is not running after start")
+        delay(200)
+        assertIs<Timer.State.NotRunning.Finished>(timer.state.value, "timer was not finished after time elapsed")
+        timer.start()
+        assertIs<Timer.State.NotRunning.Finished>(timer.state.value, "was able to start timer after finish")
+        timer.pause()
+        assertIs<Timer.State.NotRunning.Finished>(timer.state.value, "was able to pause timer after finish")
+    }
+
+    @Test
+    fun stateTransitions() {
+        checkStateTransitions(TickCorrection.Offset(-Duration.milliseconds(5)))
+        checkStateTransitions(TickCorrection.Adaptive)
+    }
+
+    @Test
+    fun awaitFinish(): Unit = runBlocking {
+        val timer = TimerProvider.monotonic(
+            duration = Duration.milliseconds(100),
+            coroutineScope = this,
+            coroutineContext = this.coroutineContext
+        )
+
+        withTimeout(Duration.milliseconds(200)) {
+            timer.start()
+            timer.awaitFinish()
+        }
+    }
+
+    private fun checkElapsedFlow(tickCorrection: TickCorrection): Unit = runBlocking {
```

Comment:
@avdyushin you suggested using `BaseFlowTest` for the test, however in my opinion the setup is more complicated with minimal benefits. could you share what you had in mind, maybe I missed something? 

the complexity is caused by the fact I need to compare the batches of the flow output and capturing using `test` is tricky

Reply:
Setup is straight forward:
```
open class TimerStateFlowTest : SimpleFlowTest<Timer.State>() {
    protected val timer = Timer()
    override val flow = suspend { timer }
}

class StartTimerTests : TimerStateFlowTest() {
    fun testTime = testWithFlow {
        action {
             timer.start()
        }
        test {
             assertIs<Timer.State.Running>(it)
        }
    }
}
```

But for collecting batches indeed is tricky. Let keep it as is or you can improve TestFlow to make easy collect states list, up to you.